### PR TITLE
Refactor/break kzg srs

### PR DESCRIPTION
**Pull request by Tabaie**
The KZG SRS breaks up naturally into a prover component (G1[:]) and a verifier component (G1[0], G2[0:2]), which only have one G1 point in common (which is usually standardized and can be omitted anyway).
Treating it as a single object results in very long Plonk verifying keys for example, which can be problematic in Smart Contracts.

This PR implements the break-up.

WARNING: I've removed pretty much all pass-by-pointers I encountered. In case some of that was performance-critical, lmk and I'll add it back in.

### CODE REVIEW

File: internal/generator/kzg/template/marshal.go.tmpl [hunk 1]
```diff
@@ -4,15 +4,62 @@ import (
...
+	for _, v := range toEncode {
+		if err := enc.Encode(v); err != nil {
+			return enc.BytesWritten(), err
+		}
+	}
+
+	return enc.BytesWritten(), nil
+}
+
+// WriteRawTo writes binary encoding of Proof to w without point compression
```
Comment (gbotrel) [inline internal/generator/kzg/template/marshal.go.tmpl:25]: Proof -> VerifyingKey

File: internal/generator/kzg/template/marshal.go.tmpl [hunk 2]
```diff
@@ -4,15 +4,62 @@ import (
...
+		}
+	}
+
+	return enc.BytesWritten(), nil
+}
+
+// WriteTo writes binary encoding of the entire SRS
 func (srs *SRS) WriteTo(w io.Writer) (int64, error) {
-	// encode the SRS
+	// encode the VerifyingKey
```
Comment (gbotrel) [inline internal/generator/kzg/template/marshal.go.tmpl:56]: bad comment.

Here I would rather call srs.Pk.WriteTo, srs.Vk.WriteTo, easier to maintain for future changes.

There is a duplicate point but it's not a crazy overhead.

File: internal/generator/kzg/template/marshal.go.tmpl [hunk 3]
```diff
@@ -24,15 +71,53 @@ func (srs *SRS) WriteTo(w io.Writer) (int64, error) {
 	return enc.BytesWritten(), nil
 }
 
+// ReadFrom decodes ProvingKey data from reader.
+func (pk *ProvingKey) ReadFrom(r io.Reader) (int64, error) {
+	// decode the ProvingKey
+	dec := {{ .CurvePackage }}.NewDecoder(r)
+
+	toDecode := []interface{}{
+		&pk.G1,
```
Comment (gbotrel) [inline internal/generator/kzg/template/marshal.go.tmpl:80]: just `dec.Decode(&pk.G1)`

File: internal/generator/kzg/template/marshal.go.tmpl [hunk 4]
```diff
@@ -24,15 +71,53 @@ func (srs *SRS) WriteTo(w io.Writer) (int64, error) {
...
 func (srs *SRS) ReadFrom(r io.Reader) (int64, error) {
-	// decode the SRS
+	// decode the VerifyingKey
 	dec := {{ .CurvePackage }}.NewDecoder(r)
 
 	toDecode := []interface{}{
-		&srs.G2[0],
-		&srs.G2[1],
-		&srs.G1,
+		&srs.Vk.G2[0],
```
Comment (gbotrel) [inline internal/generator/kzg/template/marshal.go.tmpl:118]: same `srs.Pk.ReadFrom...` `srs.Vk.ReadFrom(...)`

**Review by gbotrel — CHANGES_REQUESTED** [review]

**Comment by ThomasPiellard** [comment]
Could you modify 
```
        _, err := foldedQuotients.MultiExp(quotients, randomNumbers, config)
	if err != nil {
		return nil
	}
```

to 
```
        _, err := foldedQuotients.MultiExp(quotients, randomNumbers, config)
 	if err != nil {
		return err
	}
```

in BatchVerifyMultiPoints in the kzg.go files (e.g. l.419 for bls377) ? (Spotted by @kevaundray )

File: internal/generator/kzg/template/kzg.go.tmpl [hunk 5]
```diff
@@ -23,12 +23,22 @@ var (
 // Digest commitment of a polynomial.
 type Digest = {{ .CurvePackage }}.G1Affine
 
-// SRS stores the result of the MPC
-type SRS struct {
+// Proving and Verifying keys together constitute the SRS (result of the MPC)
+type ProvingKey struct {
```
Comment (gbotrel) [inline internal/generator/kzg/template/kzg.go.tmpl:27]: godoc format: struct doc should be // ProvingKey ...

(else it isn't render properly in godoc)

**Review by gbotrel — APPROVED** [review]
minor comment on godoc format but lgtm 👍

**Review by ThomasPiellard — APPROVED** [review]

**Pull request merged**